clang/lib/ABI/QualTypeMapper.cpp

```diff
@@ -0,0 +1,209 @@
+                                  getTypeAlign(QualType(BT, 0)), false);
+
+  case BuiltinType::Char_S:
+  case BuiltinType::SChar:
+  case BuiltinType::Short:
+    return Builder.getIntegerType(ASTCtx.getCharWidth(),
+                                  getTypeAlign(QualType(BT, 0)), true);
+
+  case BuiltinType::WChar_U:
+    return Builder.getIntegerType(ASTCtx.getCharWidth(),
+                                  getTypeAlign(QualType(BT, 0)), false);
+
+  case BuiltinType::WChar_S:
+    return Builder.getIntegerType(ASTCtx.getCharWidth(),
+                                  getTypeAlign(QualType(BT, 0)), true);
+
+  case BuiltinType::Char8:
+    return Builder.getIntegerType(8, getTypeAlign(QualType(BT, 0)), false);
+
+  case BuiltinType::Char16:
+    return Builder.getIntegerType(16, getTypeAlign(QualType(BT, 0)), false);
+
+  case BuiltinType::Char32:
+    return Builder.getIntegerType(32, getTypeAlign(QualType(BT, 0)), false);
+
+  case BuiltinType::Int:
+  case BuiltinType::UInt:
+    return Builder.getIntegerType(ASTCtx.getIntWidth(QualType(BT, 0)),
+                                  getTypeAlign(QualType(BT, 0)),
+                                  BT->getKind() == BuiltinType::Int);
+
+  case BuiltinType::Long:
+  case BuiltinType::ULong:
+    return Builder.getIntegerType(ASTCtx.getTypeSize(QualType(BT, 0)),
+                                  getTypeAlign(QualType(BT, 0)),
+                                  BT->getKind() == BuiltinType::Long);
+
+  case BuiltinType::LongLong:
+  case BuiltinType::ULongLong:
+    return Builder.getIntegerType(ASTCtx.getTypeSize(QualType(BT, 0)),
+                                  getTypeAlign(QualType(BT, 0)),
+                                  BT->getKind() == BuiltinType::LongLong);
+
+  case BuiltinType::Int128:
+  case BuiltinType::UInt128:
+    return Builder.getIntegerType(128, getTypeAlign(QualType(BT, 0)),
+                                  BT->getKind() == BuiltinType::Int128);
+
+  case BuiltinType::Half:
+  case BuiltinType::Float16:
+    return Builder.getFloatType(llvm::APFloat::IEEEhalf(),
+                                getTypeAlign(QualType(BT, 0)));
+
+  case BuiltinType::Float:
+    return Builder.getFloatType(llvm::APFloat::IEEEsingle(),
+                                getTypeAlign(QualType(BT, 0)));
+
+  case BuiltinType::Double:
+    return Builder.getFloatType(llvm::APFloat::IEEEdouble(),
+                                getTypeAlign(QualType(BT, 0)));
+
+  case BuiltinType::LongDouble:
+    return Builder.getFloatType(ASTCtx.getFloatTypeSemantics(QualType(BT, 0)),
+                                getTypeAlign(QualType(BT, 0)));
+
+  case BuiltinType::BFloat16:
+    return Builder.getFloatType(llvm::APFloat::BFloat(),
+                                getTypeAlign(QualType(BT, 0)));
+
+  case BuiltinType::Float128:
+    return Builder.getFloatType(llvm::APFloat::IEEEquad(),
+                                getTypeAlign(QualType(BT, 0)));
```

Comment:
You can use `getFloatTypeSemantics()` instead for the FP types, it will correctly determine what these mean per target.